core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -818,19 +828,37 @@ protected boolean shouldCommitDuringPrepare(PrepareCommand command, TxInvocation
              command.isOnePhaseCommit();
    }
 
-   protected final void wrapEntriesForPrepare(TxInvocationContext ctx, PrepareCommand command) throws Throwable {
+   protected final InvocationStage wrapEntriesForPrepare(TxInvocationContext ctx, PrepareCommand command) throws Throwable {
       if (!ctx.isOriginLocal() || command.isReplayEntryWrapping()) {
-         for (WriteCommand c : command.getModifications()) {
-            c.setTopologyId(command.getTopologyId());
-            // TODO: we need async invocation since the interceptors may do remote get
-            InvocationStage visitorStage = makeStage(c.acceptVisitor(ctx, entryWrappingVisitor));
-            // Wait for the sub-command to finish. If there was an exception, rethrow it.
+         return applyModifications(ctx, command, command.getModifications(), 0);
+      }
+      return null;
+   }
+
+   private InvocationStage applyModifications(TxInvocationContext ctx, PrepareCommand command, WriteCommand[] modifications, int startIndex) throws Throwable {
+      // We need to execute modifications for the same key sequentially. In theory we could optimize
+      // this loop if there are multiple remote invocations but since remote invocations are rare, we omit this.
+      for (int i = startIndex; i < modifications.length; i++) {
+         WriteCommand c = modifications[i];
+         c.setTopologyId(command.getTopologyId());
+         InvocationStage visitorStage = makeStage(c.acceptVisitor(ctx, entryWrappingVisitor));
+
+         if (visitorStage.isDone()) {
```

Comment:
Oops, I missed both my own comments and your replies when looking at this again.